esmvaltool/diag_scripts/ipcc_ar6/corr_pattern.ncl

```diff
@@ -0,0 +1,200 @@
+  ref_ind = ind(datasetnames.eq.ref_model)
+  if isatt(variable_info[0], "alternative_dataset") then
+    alt_ref = True
+    aref_ind = ind(datasetnames.eq.variable_info[0]@alternative_dataset)
+  else
+    alt_ref = False
+  end if
+
+end
+
+begin
+  ; Output directories
+  ncdf_dir = config_user_info@work_dir + "pattern_cor.nc"
+  modproj_dir = config_user_info@work_dir + "modprojnames.txt"
+  system("mkdir -p " + config_user_info@work_dir)
+end
+
+begin
+; -----------------------------------------------------------------------------
+; -------------------- Compute correlation for one var ------------------------
+; -----------------------------------------------------------------------------
+
+  ; Save list of preproc files for provenance in collect.ncl
+  preproc_files = metadata_att_as_array(info_items, "filename")
+
+  ; Reference model
+  mod_idx = ispan(0, dim_MOD - 1, 1)
+  mod_ind_woref = mod_idx(ind(mod_idx.ne.ref_ind))
+  delete(mod_idx)
+
+  ; Make output array
+  all_cor = new((/dim_MOD-1/), float)
+  all_cor@_FillValue = 1.e20
+  all_cor!0 = "models"
+  all_cor&models = datasetnames(mod_ind_woref)
+  ; Pass on alt models
+  if alt_ref then
+    all_cor@alt_obs =  variable_info[0]@alternative_dataset
+  else
+    all_cor@alt_obs = "none"
+  end if
+
+  ; Loop over models, with ref model processed first
+  model_ind = array_append_record(ref_ind, mod_ind_woref, 0)
+  do iloop = 0, dim_MOD - 1
+    imod = model_ind(iloop)
+    log_info("Processing " + datasetnames(imod))
+
+    ; Extract model corresponding data
+    var = read_data(info_items[imod])
+
+    ; Calculate annual mean
+    data_yearly = time_operations(var, \
+                                  toint(info_items[imod]@start_year), \
+                                  toint(info_items[imod]@end_year), \
+                                  "average", "yearly", True)
+
+    ; Mean over the years
+    mean_years = dim_avg_n_Wrap(data_yearly, 0)
+    delete(data_yearly)
+
+    ; Compute centred pattern correlation (ref_model will be first so no error)
+    if datasetnames(imod).eq.ref_model then
+      data_ref = mean_years
+    else
+      ; Since ref model processed first, move all models up one spot
+      all_cor(iloop-1) = pattern_cor(data_ref, mean_years, 1.0, 0)
+    end if
+    delete(mean_years)
+    delete(var)
+  end do
+
+  ; *************************************
+  ; calculate means over ensemble members
+  ; *************************************
+
+  datasetnames := datasetnames(mod_ind_woref)
+  projectnames := projectnames(mod_ind_woref)
+
+  ndim := dimsizes(datasetnames)
+  ensemble_assign = new(ndim, integer)
+
+  enum = 1
+  temp = datasetnames(0)
+  do gg = 0, ndim - 1
+    if (temp.eq.datasetnames(gg)) then
+      ensemble_assign(gg) = enum
+    else
+      enum = enum + 1
+      ensemble_assign(gg) = enum
+      temp = datasetnames(gg)
+    end if
+  end do
+  ensemble_assign@model = datasetnames
+  delete(temp)
```

Comment:
Actually just changing this to work no matter the order is easy, this might not be the most elegant solution but it should work, also you never use ensemble_assign@model after, so you could delete that line
```suggestion
  unique_models = get_unique_values(datasetnames)
  do gg = 0, dimsizes(unique_models) - 1
    ensemble_assign = where(datasetnames.eq.unique_models(gg), gg + 1, ensemble_assign)
  end do
  ensemble_assign@model = datasetnames
  delete(temp)
```